### PR TITLE
merge cumprod rules to fix ambig on julia 1.6

### PR DESCRIPTION
**Pull request by oxinabox**
Suggested by @devmotion  in https://github.com/JuliaDiff/ChainRules.jl/pull/781
should resolve same issue.
I am not sure which is cleaner

### CODE REVIEW

**Review by devmotion — COMMENTED** [review]
I think I'm slightly in favour of this PR, since it reduces the number of methods and `cumprod(::AbstractVector)` is forwarded to the keyword argument version: https://github.com/JuliaLang/julia/blob/63e95d47e692352fac5541b6ee835b573f4c4898/base/accumulate.jl#L227

Unfortunately, the diff is significantly larger than in #781 - and it requires a few additional changes to the tests since there's no `rrule(cumprod, ::AbstractVector)` anymore (so one has to set `dims=1` explicitly in these tests).

File: src/rulesets/Base/mapreduce.jl [hunk 1]
```diff
@@ -325,52 +325,37 @@ end
...
-            else
-                dy
-            end)
-            )
-        return (NoTangent(), dx_thunk)
-    end
-    return y, cumprod_pullback_1
-end
-
 function rrule(::typeof(cumprod), x::AbstractArray{<:Real}; dims::Integer)
```
Comment (oxinabox) [inline src/rulesets/Base/mapreduce.jl:328]:
```suggestion
function rrule(::typeof(cumprod), x::AbstractArray{<:Real}; dims::Integer=1)
```

File: src/rulesets/Base/mapreduce.jl [hunk 2]
```diff
@@ -325,52 +325,37 @@ end
...
-                dy
-            end)
-            )
-        return (NoTangent(), dx_thunk)
-    end
-    return y, cumprod_pullback_1
-end
-
-function rrule(::typeof(cumprod), x::AbstractArray{<:Real}; dims::Integer)
+function rrule(::typeof(cumprod), x::AbstractArray{<:Real}; dims::Integer=1)
```
Comment (devmotion) [inline src/rulesets/Base/mapreduce.jl:328]: Is it OK to put a default value for `dims` even though there's no default value in base? https://github.com/JuliaLang/julia/blob/63e95d47e692352fac5541b6ee835b573f4c4898/base/accumulate.jl#L193

**Comment by oxinabox** [comment]
its actually fine for the `rrule` to always default dims to 1. (So I have made that change)
Since:
 - in the vector case this is what is wanted and is the correct default
 - in the nonvector case then a keyword argument is required by the primal, so the default will never be hit.

File: src/rulesets/Base/mapreduce.jl (hunk 2 above)
Comment (oxinabox) [reply src/rulesets/Base/mapreduce.jl:328]: yes, it is see:
https://github.com/JuliaDiff/ChainRules.jl/pull/784#issuecomment-1936107183

**Comment by mcabbott** [comment]
> in the nonvector case then a keyword argument is required by the primal, so the default will never be hit.

What does this mean? The primal computation is done inside the rrule, and does use the default. 

I think that will mean `gradient(x -> sum(abs, cumprod(x)), [1 2; 3 4.])` is not an error, whereas without gradient it would be.

That said, I have no memory of why there are two rrules. Besides this issue there might have been performance concerns?

**Comment by oxinabox** [comment]
> I think that will mean gradient(x -> sum(abs, cumprod(x)), [1 2; 3 4.]) is not an error, whereas without gradient it would be.

That is a point.
I had been thinking we could rely on the primal computation being correct.
But I guess we can not.

We will have to do somehting more complex like set the rrule as `dims=nothing`
then if `ndims(x)==1` set it to `dims=1`

**Comment by devmotion** [comment]
> We will have to do somehting more complex like set the rrule as dims=nothing

Can't we just omit the default value and adjust the tests?

File: src/rulesets/Base/mapreduce.jl (hunk 2 above)
Comment (oxinabox) [inline src/rulesets/Base/mapreduce.jl:328]:
```suggestion
function rrule(::typeof(cumprod), x::AbstractArray{<:Real}; dims::Integer=nothing)
    if isnothing(dims) && ndims(x)==1
        dims=1
    else
        throw(UndefKeywordError(:dims)
    end

```

**Comment by mcabbott** [comment]
My vote is that https://github.com/JuliaDiff/ChainRules.jl/pull/781 is much simpler. 

Besides worrying about error cases, I think this kind of change would want to be checked very carefully for type stability or performance regressions.

**Pull request closed without merging**